README.md
# Lockit MongoDB adapter

[![Build Status](https://travis-ci.org/zemirco/lockit-mongodb-adapter.svg?branch=master)](https://travis-ci.org/zemirco/lockit-mongodb-adapter) [![NPM version](https://badge.fury.io/js/lockit-mongodb-adapter.svg)](http://badge.fury.io/js/lockit-mongodb-adapter)

MongoDB adapter for [Lockit](https://github.com/zemirco/lockit).

## Installation

`npm install lockit-mongodb-adapter`

```js
var adapter = require('lockit-mongodb-adapter');
```

## Configuration

The following setting is required.

```js
exports.db = {
  url: 'mongodb://127.0.0.1/',
  name: 'test',
  collection: 'users'
};
```

## Features

### 1. Create user

`adapter.save(name, email, pass, callback)`

 - `name`: String - i.e. 'john'
 - `email`: String - i.e. 'john@email.com'
 - `pass`: String - i.e. 'password123'
 - `callback`: Function - `callback(err, user)` where `user` is the new user now in our database.

The `user` object has the following properties

 - `name`: username chosen during sign up
 - `email`: email that was provided at the beginning
 - `signupTimestamp`: Date object to remember when the user signed up
 - `signupToken`: unique token sent to user's email for email verification
 - `signupTokenExpires`: Date object usually 24h ahead of `signupTimestamp`
 - `failedLoginAttempts`: save failed login attempts during login process, default is `0`
 - `salt`: salt generated by `crypto.randomBytes()`
 - `derived_key`: password hash generated by pbkdf2
 - `_id`: document id

```js
adapter.save('john', 'john@email.com', 'secret', function(err, user) {
  if (err) console.log(err);
  console.log(user);
  // {
  //  name: 'john',
  //  email: 'john@email.com',
  //  signupToken: 'ef32a95a-d6ee-405a-8e4b-515b235f7c54',
  //  signupTimestamp: Wed Jan 15 2014 19:08:27 GMT+0100 (CET),
  //  signupTokenExpires: Wed Jan 15 2014 19:08:27 GMT+0100 (CET),
  //  failedLoginAttempts: 0,
  //  salt: '48cf9da376703199c30ba5c274580c98',
  //  derived_key: '502967e5a6e55091f4c2c80e7989623f051070fd',
  //  _id: 52d6ce9b651b4d825351641f
  // }
});
```

### 2. Find user

`adapter.find(match, query, callback)`

 - `match`: String - one of the following: 'name', 'email' or 'signupToken'
 - `query`: String - corresponds to `match`, i.e. 'john@email.com'
 - `callback`:  Function - `callback(err, user)`

```js
adapter.find('name', 'john', function(err, user) {
  if (err) console.log(err);
  console.log(user);
  // {
  //   name: 'john',
  //   email: 'john@email.com',
  //   signupToken: '3a7f0f54-32f0-44f7-97c6-f1470b94c170',
  //   signupTimestamp: Fri Apr 11 2014 21:31:54 GMT+0200 (CEST),
  //   signupTokenExpires: Sat Apr 12 2014 21:31:54 GMT+0200 (CEST),
  //   failedLoginAttempts: 0,
  //   salt: '753981e8d8e30e8047cf5685d1f0a0d4',
  //   derived_key: '18ce03eddab6729aeaaf76729c90cb31f16a863c',
  //   _id: 5348432a98a8a6a4fef1f595
  // }
});
```

### 3. Update user

`adapter.update(user, callback)`

 - `user`: Object - must have `_id` and `_rev` properties
 - `callback`: Function - `callback(err, user)` - `user` is the updated user object

```js
// get a user from db first
adapter.find('name', 'john', function(err, user) {
  if (err) console.log(err);

  // add some new properties to our existing user
  user.newKey = 'and some value';
  user.hasBeenUpdated = true;

  // save updated user to db
  adapter.update(user, function(err, user) {
    if (err) console.log(err);
    // ...
  });
});
```

### 4. Remove user

`adapter.remove(name, callback)`

 - `name`: String
 - `callback`: Function - `callback(err, res)` - `res` is `true` if everything went fine

```js
adapter.remove('john', function(err, res) {
  if (err) console.log(err);
  console.log(res);
  // true
});
```

## Test

`grunt`

## License

MIT


History.md

##### 0.4.3 / 2015-07-03

- use eslint
- update dependencies
- use make instead grunt

##### 0.4.2 / 2014-07-31

- use `findOne()` instead `find().nextObject()`

##### 0.4.1 / 2014-05-22

- code refactoring
- update dependencies

##### 0.4.0 / 2014-04-17

- database config has to be an Object and the database name isn't part of the
  connection string anymore

  **old**

  ```js
  exports.db = 'mongodb://127.0.0.1/test';
  exports.dbCollection = 'users';
  ```

  **new**

  ```js
  exports.db = {
    url: 'mongodb://127.0.0.1/',
    name: 'test',
    collection: 'users'
  };
  ```

##### 0.3.0 / 2014-04-11

- key `username` is now `name`
- use built-in [pbkdf2](http://nodejs.org/api/crypto.html#crypto_crypto_pbkdf2_password_salt_iterations_keylen_callback)
  instead [bcrypt](https://github.com/ncb000gt/node.bcrypt.js/)

##### 0.2.0 / 2014-04-03

- simplify `remove()` method

  ```js
  adapter.remove(match, query, callback)
  ```

  becomes

  ```js
  adapter.remove(username, callback)
  ```

...

##### 0.1.0 / 2013-01-22

 - drop `dbUrl` and use `db` instead
 - use new `config.js` structure


test/test.js
'use strict';

var should = require('should');
var config = require('./config.js');

var adapter = require('../index.js')(config);

// wait for connection to db
before(function(done) {

  setTimeout(function() {
    done();
  }, 1000);

});

describe('lockit-mongodb-adapter', function() {

  // needed later for test 'find user by signup token'
  var _tmp_signupToken = '';

  it('should create a new user', function(done) {
    adapter.save('john', 'john@email.com', 'secret', function(err, res) {
      if (err) {console.log(err); }
      res.should.have.property('signupToken');
      res.signupToken.should.match(/[0-9a-f]{22}|[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}/);
      res.should.have.property('failedLoginAttempts');
      res.failedLoginAttempts.should.equal(0);
      _tmp_signupToken = res.signupToken;
      res.email.should.equal('john@email.com');
      done();
    });
  });

  it('should find a user by name', function(done) {
    adapter.find('name', 'john', function(err, res) {
      if (err) {console.log(err); }
      res.name.should.equal('john');
      res.email.should.equal('john@email.com');
      done();
    });
  });

  it('should return undefined when no user is found', function(done) {
    adapter.find('name', 'jim', function(err, res) {
      if (err) {console.log(err); }
      should.not.exist(err);
      should.not.exist(res);
      done();
    });
  });

  it('should find a user by email', function(done) {
    adapter.find('email', 'john@email.com', function(err, res) {
      if (err) {console.log(err); }
      res.name.should.equal('john');
      res.email.should.equal('john@email.com');
      done();
    });
  });

  it('should find a user by signup token', function(done) {
    adapter.find('signupToken', _tmp_signupToken, function(err, res) {
      if (err) {console.log(err); }
      res.name.should.equal('john');
      res.email.should.equal('john@email.com');
      done();
    });
  });

  it('should update an existing user', function(done) {
    adapter.find('name', 'john', function(err, doc) {
      if (err) {console.log(err); }
      doc.test = 'works';
      doc.editet = true;
      adapter.update(doc, function(updateErr, res) {
        if (updateErr) {console.log(updateErr); }
        res.test.should.equal('works');
        res.editet.should.be.true;
        done();
      });
    });
  });

  it('should remove a user', function(done) {
    adapter.save('jeff', 'jeff@email.com', 'secret', function(err) {
      if (err) {console.log(err); }
      adapter.remove('jeff', function(removeErr, res) {
        if (removeErr) {console.log(removeErr); }
        res.should.be.true;
        done();
      });
    });
  });

  it('should return an error when remove cannot find a user', function(done) {
    adapter.remove('steve', function(err) {
      err.message.should.equal('lockit - Cannot find user "steve"');
      done();
    });
  });

});

// remove users db
after(function(done) {
  adapter.remove('john', done);
});


test/config.js

exports.db = {
  url: 'mongodb://127.0.0.1/',
  name: 'test',
  collection: 'users'
};

exports.signup = {
  tokenExpiration: '1 day'
};


index.js
'use strict';

var MongoClient = require('mongodb').MongoClient;
var uuid = require('node-uuid');
var pwd = require('couch-pwd');
var ms = require('ms');
var moment = require('moment');



/**
 * Adapter constructor function.
 *
 * @example
   var Adapter = require('lockit-mongodb-adapter');
   var config = require('./config.js');
   var adapter = new Adapter(config);
 *
 * @param {Object} config - Lockit configuration
 * @constructor
 */
var Adapter = module.exports = function(config) {

  if (!(this instanceof Adapter)) {return new Adapter(config); }

  this.config = config;
  this.collection = config.db.collection;

  // create connection string
  var url = config.db.url + config.db.name;

  // create connection as soon as module is required and share global db object
  var that = this;
  MongoClient.connect(url, function(err, database) {
    if (err) {throw err; }
    that.db = database;
  });

};



/**
 * Create new user.
 *
 * @example
   adapter.save('john', 'john@email.com', 'secret', function(err, user) {
     if (err) console.log(err);
     console.log(user);
     // {
     //  name: 'john',
     //  email: 'john@email.com',
     //  signupToken: 'ef32a95a-d6ee-405a-8e4b-515b235f7c54',
     //  signupTimestamp: Wed Jan 15 2014 19:08:27 GMT+0100 (CET),
     //  signupTokenExpires: Wed Jan 15 2014 19:08:27 GMT+0100 (CET),
     //  failedLoginAttempts: 0,
     //  salt: '48cf9da376703199c30ba5c274580c98',
     //  derived_key: '502967e5a6e55091f4c2c80e7989623f051070fd',
     //  _id: 52d6ce9b651b4d825351641f
     // }
   });
 *
 * @param {String} name - User name
 * @param {String} email - User email
 * @param {String} pw - Plain text user password
 * @param {Function} done - Callback function `function(err, user){}`
 */
Adapter.prototype.save = function(name, email, pw, done) {
  var that = this;

  var now = moment().toDate();
  var timespan = ms(that.config.signup.tokenExpiration);
  var future = moment().add(timespan, 'ms').toDate();

  var user = {
    name: name,
    email: email,
    signupToken: uuid.v4(),
    signupTimestamp: now,
    signupTokenExpires: future,
    failedLoginAttempts: 0
  };

  // create salt and hash
  pwd.hash(pw, function(err, salt, hash) {
    if (err) {return done(err); }
    user.salt = salt;
    user.derived_key = hash;
    that.db.collection(that.collection).save(user, function(saveErr, result) {
      done(saveErr, result.ops[0]);
    });
  });
};



/**
 * Find user. Match is either `'name'`, `'email'` or `'signupToken'`.
 *
 * @example
   adapter.find('name', 'john', function(err, user) {
     if (err) console.log(err);
     console.log(user);
     // {
     //   name: 'john',
     //   email: 'john@email.com',
     //   signupToken: '3a7f0f54-32f0-44f7-97c6-f1470b94c170',
     //   signupTimestamp: Fri Apr 11 2014 21:31:54 GMT+0200 (CEST),
     //   signupTokenExpires: Sat Apr 12 2014 21:31:54 GMT+0200 (CEST),
     //   failedLoginAttempts: 0,
     //   salt: '753981e8d8e30e8047cf5685d1f0a0d4',
     //   derived_key: '18ce03eddab6729aeaaf76729c90cb31f16a863c',
     //   _id: 5348432a98a8a6a4fef1f595
     // }
   });
 *
 * @param {String} match - Property to find user by. `'name'`, `'email'` or `'signupToken'`
 * @param {String} query - Corresponding value to `match`
 * @param {Function} done - Callback function `function(err, user){}`
 */
Adapter.prototype.find = function(match, query, done) {
  var qry = {};
  qry[match] = query;
  this.db.collection(this.collection).findOne(qry, done);
};



/**
 * Update existing user.
 *
 * @example
   // get user from db
   adapter.find('name', 'john', function(err, user) {
     if (err) console.log(err);

     // add some new properties
     user.newKey = 'and some value';
     user.hasBeenUpdated = true;

     // save updated user to db
     adapter.update(user, function(err, user) {
       if (err) console.log(err);
       // ...
     });
   });
 *
 * @param {Object} user - Existing user from db
 * @param {Function} done - Callback function `function(err, user){}`
 */
Adapter.prototype.update = function(user, done) {
  var that = this;
  // update user in db
  that.db.collection(that.collection).save(user, function(err) {
    if (err) {return done(err); }
    // res is not the updated user object! -> find manually
    that.db.collection(that.collection).findOne({_id: user._id}, done);
  });
};



/**
 * Delete existing user.
 *
 * @example
   adapter.remove('john', function(err, res) {
     if (err) console.log(err);
     console.log(res);
     // true
   });
 *
 * @param {String} name - User name
 * @param {Function} done - Callback function `function(err, res){}`
 */
Adapter.prototype.remove = function(name, done) {
  this.db.collection(this.collection).remove({name: name}, function(err, result) {
    if (err) {return done(err); }
    if (result.result.n === 0) {return done(new Error('lockit - Cannot find user "' + name + '"')); }
    done(null, true);
  });
};
